File: source/com/janilla/acmedashboard/AcmeDashboard.js
/*
 * MIT License
 *
 * Copyright (c) 2024 Diego Schivo
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all
 * copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 */
import Component from "./Component.js";
import Customers from "./Customers.js";
import Home from "./Home.js";
import Dashboard, { Layout } from "./Dashboard.js";
import Invoice from "./Invoice.js";
import Invoices from "./Invoices.js";
import Login from "./Login.js";
import RenderEngine from "./RenderEngine.js";
import heroIcons from "./heroIcons.js";

const pages1 = {
	"/": () => new Home(),
	"/dashboard": () => new Dashboard(),
	"/dashboard/customers": () => new Customers(),
	"/dashboard/invoices": () => new Invoices(),
	"/dashboard/invoices/create": () => new Invoice(),
	"/login": () => new Login(),
};

const pages2 = new Map();
pages2.set(new RegExp("^/dashboard/invoices/(\\d+)/edit$"), x => new Invoice(parseInt(x, 10)));

const currencyFormatter = new Intl.NumberFormat("en-US", {
	style: "currency",
	currency: "USD"
});

const dateFormatter = new Intl.DateTimeFormat("en-US", {
	day: "numeric",
	month: "short",
	year: "numeric",
});

export default class AcmeDashboard extends Component {

	icons = heroIcons;

	logo = new Logo();

	page;

	layout = new Layout();

	constructor() {
		super("AcmeDashboard");
	}

	get content() {
		return this.page instanceof Home || this.page instanceof Login ? this.page : this.layout;
	}

	async run() {
		const u = new URL(location.href);
		let p = u.pathname;
		if (p === "/")
			p = await (await fetch("/api/authentication")).json() ? "/dashboard" : "/";
		this.page = this.createPage(p);
		document.title = [this.page.title, document.title.split(" | ").at(-1)].filter(x => x).join(" | ");
		const re = new RenderEngine(document.body.querySelectorAll("template"));
		document.body.innerHTML = "";
		document.body.append(...re.render({ value: this }));
		this.listen();
		history.replaceState(this.page.state, "", p + u.search);
		document.addEventListener("click", this.handleClick);
		addEventListener("urlchange", this.handleUrlchange);
		addEventListener("popstate", this.handlePopstate);
	}

	tryRender(re) {
		const o = re.stack.at(-1);
		if (typeof o.key === "string" && (o.key === "icon" || o.key.endsWith("Icon"))) {
			o.value = heroIcons[o.value];
			return true;
		}
		if (typeof o.key === "string" && (o.key === "amount" || o.key.endsWith("Amount"))) {
			o.value = currencyFormatter.format(o.value);
			return true;
		}
		if (typeof o.key === "string" && (o.key === "date" || o.key.endsWith("Date"))) {
			o.value = dateFormatter.format(new Date(o.value));
			return true;
		}
		return super.tryRender(re);
	}

	listen() {
		if (this.content === this.layout)
			this.layout.listen();
		this.page.listen();
	}

	refreshPage(p) {
		const l = this.content === this.layout;
		this.page = p;
		document.title = [this.page.title, document.title.split(" | ").at(-1)].filter(x => x).join(" | ");
		if (l && this.content === this.layout) {
			const m = document.querySelector("main");
			m.innerHTML = "";
			const nn = this.renderEngine.render({ value: this.page });
			m.append(...nn);
			this.page.renderedNodes = nn;
			this.page.listen();
		} else {
			const re = new RenderEngine(this.renderEngine.templates);
			document.body.innerHTML = "";
			document.body.append(...re.render({ value: this }));
			this.listen();
		}
	}

	createPage(path) {
		const p = path.length > 1 && path.endsWith("/") ? path.substring(0, path.length - 1) : path;
		const q = pages1[p];
		if (q)
			return q();
		for (const [k, v] of pages2) {
			const m = p.match(k);
			if (m)
				return v(...m.slice(1));
		}
		throw new Error();
	}

	handleClick = e => {
		const a = e.target.closest("a");
		if (!a)
			return;
		e.preventDefault();
		dispatchEvent(new CustomEvent("urlchange", { detail: { url: new URL(a.href) } }));
	}

	handleUrlchange = e => {
		const u = e.detail.url;
		history.pushState({}, "", u.pathname + u.search);
		this.refreshPage(this.createPage(u.pathname));
		history.replaceState(this.page.state, "");
		dispatchEvent(new Event("popstate"));
	}

	handlePopstate = e => {
		if (!e.state)
			return;
		const p = this.createPage(document.location.pathname);
		p.state = e.state;
		this.refreshPage(p);
	}
}

class Logo extends Component {

	constructor() {
		super("AcmeDashboard");
	}

	tryRender(re) {
		return re.match([this], (_, o) => {
			o.template = "AcmeDashboard-Logo";
		});
	}
}

File: main/source/com/janilla/acmedashboard/AcmeDashboard.java
/*
 * MIT License
 *
 * Copyright (c) 2024-2025 Diego Schivo
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all
 * copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 */
package com.janilla.acmedashboard;

import java.net.InetSocketAddress;
import java.net.URI;
import java.nio.file.Files;
import java.nio.file.Path;
import java.util.Collections;
import java.util.List;
import java.util.Map;
import java.util.Properties;
import java.util.UUID;
import java.util.concurrent.atomic.AtomicReference;

import javax.net.ssl.SSLContext;

import com.janilla.http.HttpHandler;
import com.janilla.http.HttpServer;
import com.janilla.json.Json;
import com.janilla.json.MapAndType;
import com.janilla.json.ReflectionJsonIterator;
import com.janilla.net.Net;
import com.janilla.persistence.ApplicationPersistenceBuilder;
import com.janilla.persistence.Persistence;
import com.janilla.reflect.Factory;
import com.janilla.reflect.Flatten;
import com.janilla.util.Util;
import com.janilla.web.ApplicationHandlerBuilder;
import com.janilla.web.Handle;
import com.janilla.web.Render;
import com.janilla.web.RenderableFactory;
import com.janilla.web.Renderer;

public class AcmeDashboard {

	public static final AtomicReference<AcmeDashboard> INSTANCE = new AtomicReference<>();

	public static void main(String[] args) {
		try {
			var pp = new Properties();
			try (var s1 = AcmeDashboard.class.getResourceAsStream("configuration.properties")) {
				pp.load(s1);
				if (args.length > 0) {
					var p = args[0];
					if (p.startsWith("~"))
						p = System.getProperty("user.home") + p.substring(1);
					try (var s2 = Files.newInputStream(Path.of(p))) {
						pp.load(s2);
					}
				}
			}
			var x = new AcmeDashboard(pp);

			HttpServer s;
			{
				SSLContext c;
				try (var is = Net.class.getResourceAsStream("testkeys")) {
					c = Net.getSSLContext("JKS", is, "passphrase".toCharArray());
				}
				s = x.factory.create(HttpServer.class, Map.of("sslContext", c, "handler", x.handler));
			}
			var p = Integer.parseInt(x.configuration.getProperty("acme-dashboard.server.port"));
			s.serve(new InetSocketAddress(p));
		} catch (Throwable e) {
			e.printStackTrace();
		}
	}

	public Properties configuration;

	public Factory factory;

	public Persistence persistence;

	public RenderableFactory renderableFactory;

	public HttpHandler handler;

	public MapAndType.TypeResolver typeResolver;

	public Iterable<Class<?>> types;

	public AcmeDashboard(Properties configuration) {
		if (!INSTANCE.compareAndSet(null, this))
			throw new IllegalStateException();
		this.configuration = configuration;
		types = Util.getPackageClasses(getClass().getPackageName()).filter(x -> !x.getPackageName().endsWith(".test"))
				.toList();
		factory = new Factory(types, this);
		typeResolver = factory.create(MapAndType.DollarTypeResolver.class);
		{
			var p = configuration.getProperty("acme-dashboard.database.file");
			if (p.startsWith("~"))
				p = System.getProperty("user.home") + p.substring(1);
			var b = factory.create(ApplicationPersistenceBuilder.class, Map.of("databaseFile", Path.of(p)));
			persistence = b.build();
		}
		renderableFactory = new RenderableFactory();
		{
			var b = factory.create(ApplicationHandlerBuilder.class);
			handler = b.build();
		}
	}

	public AcmeDashboard application() {
		return this;
	}

	@Handle(method = "GET", path = "/")
	public Object root(CustomHttpExchange exchange) {
		return exchange.getSessionEmail() != null ? URI.create("/dashboard")
				: new Index(Collections.singletonMap("authentication", null));
	}

	@Handle(method = "GET", path = "/login")
	public Index login() {
		return new Index(Map.of());
	}

	@Handle(method = "GET", path = "/dashboard")
	public Index dashboard() {
		var x = DashboardApi.INSTANCE.get();
		return new Index(Map.of("cards", x.getCards(), "revenue", x.getRevenue(), "invoices", x.getInvoices()));
	}

	@Handle(method = "GET", path = "/dashboard/invoices")
	public Index invoices(String query, Integer page) {
		return new Index(Collections.singletonMap("invoices", InvoiceApi.INSTANCE.get().list(query, page)));
	}

	@Handle(method = "GET", path = "/dashboard/invoices/create")
	public Index createInvoice() {
		return new Index(Collections.singletonMap("invoice", new Invoice2(null, CustomerApi.INSTANCE.get().names())));
	}

	@Handle(method = "GET", path = "/dashboard/invoices/([^/]+)/edit")
	public Index editInvoice(UUID id) {
		return new Index(Collections.singletonMap("invoice",
				new Invoice2(InvoiceApi.INSTANCE.get().read(id), CustomerApi.INSTANCE.get().names())));
	}

	@Handle(method = "GET", path = "/dashboard/customers")
	public Index customers(String query) {
		return new Index(Collections.singletonMap("customers", CustomerApi.INSTANCE.get().list(query)));
	}

	@Render(template = "index.html")
	public record Index(@Render(renderer = StateRenderer.class) Map<String, Object> state) {
	}

	public static class StateRenderer<T> extends Renderer<T> {

		@Override
		public String apply(T value) {
			var x = INSTANCE.get().factory.create(ReflectionJsonIterator.class);
			x.setObject(value);
			return Json.format(x);
		}
	}

	public record Invoice2(@Flatten Invoice invoice, List<Map.Entry<String, String>> customers) {
	}
}
